### PR TITLE
Feature/python sdk

### LINKED ISSUE #100: [SDK] Python SDK Initial Implementation (v0.1.0) **Difficulty**: Advanced | **Area**: `sdks/` | **Package Name**: `urbackend`
### Description
Expand the reach of `urBackend` by creating the first official Python SDK. This will allow Python/Django/Flask developers to use urBackend as their data layer.

### Technical Requirements
- **Version**: Initial release must be `0.1.0`.
- **Package Name**: Reserve and use the name `urbackend` (or `urbackend-python` if unavailable) on PyPI.
- **Project Structure**: Follow the standard `pyproject.toml` or `setup.py` layout in a new directory: `sdks/urbackend-python`.

### Key Tasks
- Implement a `urBackendClient` class with:
  - `connect(api_key, project_id, base_url)`
  - `db.collection(name).find(query)`
  - `db.collection(name).insert(data)`
  - `auth.login(email, password)`
- **Documentation**: Provide a basic `README.md` with usage examples in Python.
- **Modeling**: Mirror the architecture and method names of the existing [TypeScript SDK](file:///e:/Majors/WEB/urBackend/sdks/urbackend-sdk) for consistency.

### Acceptance Criteria
- [ ] Successful connection to a live/local urBackend instance.
- [ ] Support for basic Auth and CRUD operations.
- [ ] Fully documented internal methods with Docstrings.
- [ ] Ready for publication to PyPI at version `0.1.0`.

### PR DESCRIPTION
**Pull request by Navyasree-ulava**
## Fixes Issue #100 

This PR adds the **official Python SDK (v0.1.0)** for urBackend, enabling Python developers to interact with urBackend APIs for authentication and database operations.

The SDK mirrors the TypeScript SDK structure while keeping the implementation simple and easy to use. It includes full support for auth, CRUD operations, token handling, and error management.

Additionally, I added comprehensive testing (`test_sdk.py`) covering real API usage, negative scenarios, and edge cases.

---

## Type of Change

* [ ] 🐛 Bug fix
* [x] ✨ New feature
* [ ] 💥 Breaking change
* [ ] 📝 Documentation update
* [ ] 🎨 UI/UX improvement
* [ ] ⚙️ Refactor / Chore

---

### Backend Verification:

* [ ] I have run `npm test` in the `backend/` directory
* [x] I have verified the API endpoints using Postman/Thunder Client
* [x] New unit tests / test script added (`test_sdk.py`)

### Frontend Verification:

* [x] Not applicable

---

### Client

* `connect(api_key, project_id, base_url)`

### Auth

* `login`, `me`, `logout`
* Automatic token storage and reuse

### Database

* `collection(name)`
* `find`, `insert`, `get_one`, `patch`, `delete`, `count`
* Supports query params (filter, sort, limit)

### Internal

* Centralized HTTP request handler
* Automatic headers (`x-api-key`, `Authorization`)
* Structured error handling (AuthError, ValidationError, NotFoundError)

---

## Testing Coverage

Tested against live API:
`https://api.ub.bitbros.in`

Covers:

* Auth flow (signup, login, logout)
* Token auto-storage and reuse
* Full CRUD operations
* Query parameters (limit, sort, filter)
* Negative cases:

  * Missing token (401)
  * Invalid login
  * Invalid collection (404)
* Edge cases:

  * Empty insert
  * Invalid IDs
  * Empty queries
* Token override behavior

All tests pass via `python test_sdk.py`.

---

## ✅ Checklist

* [x] My code follows the code style of this project
* [x] I have performed a self-review of my code
* [x] I have commented my code where necessary
* [x] My changes generate no new warnings or errors
* [x] I have updated documentation (README)

---

Built with ❤️ for **urBackend**.

## Release Notes

* **New Features**
  * Launched Python SDK for urBackend with complete authentication (sign up, login, logout, token management).
  * Added database module supporting CRUD operations, filtering, sorting, and pagination.
  * Introduced file storage module for uploads and deletions.
  * Included email sending capabilities with template support.
  * Enabled social OAuth integration.

* **Documentation**
  * Added comprehensive Python SDK README with usage examples and API reference.
  * Updated main README with Python installation and quick-start example.

* **Tests**
  * Added full test suite covering authentication, database, storage, HTTP layer, and exception handling.
  * Included live integration test script for real-world validation.

* **Chores**
  * Created project configuration and environment setup files.

### CODE REVIEW

**Comment by yash-pouranik** [comment]
@Navyasree-ulava 
Please resolve the comments added by coderabbit
Only then I am able to review manually.

**Comment by yash-pouranik** [comment]
@Navyasree-ulava 
Please rename this folder from 
sdks/urbackend-python/urbackend
 to 
sdks/urbackend-python/src

File: sdks/urbackend-python/tests/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+# urbackend-python test suite
```
Comment (yash-pouranik) [inline sdks/urbackend-python/tests/__init__.py:1]: What is this file??
Actally I am not too much in Python can u explain?? 
@Navyasree-ulava

File: sdks/urbackend-python/tests/test_auth.py
Comment (yash-pouranik) [inline sdks/urbackend-python/tests/test_auth.py:1]: Please Share results of all the tests as Screenshots.
@Navyasree-ulava

File: sdks/urbackend-python/tests/__init__.py (hunk 1 above)
Comment (Navyasree-ulava) [reply sdks/urbackend-python/tests/__init__.py:1]: This file marks the `tests/` directory as a Python package.

It helps Python and tools like pytest correctly discover and import test modules. It doesn’t contain any logic and is a common convention in Python projects for organizing test suites.

File: sdks/urbackend-python/src/client.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    RateLimitError,
+    StorageError,
+    UrBackendError,
+    ValidationError,
+)
+from .http import UrBackendHTTP
+from .mail import MailModule
+from .storage import StorageModule
+
+_DEFAULT_BASE_URL = "https://api.ub.bitbros.in"
```
Comment (yash-pouranik) [inline sdks/urbackend-python/src/client.py]: If this is the syntac to use urbackend sdk.
`client = UrBackendClient(api_key="pk_live_YOUR_KEY")`
Then by default reqs will be sent to - https://api.ub.bitbros.in/
What if a user/dev has self hosted urbackend and has their own url - suppose - api.ub.devDomain.com
then how will he be able to use the sdk??

So, Qn is. Is there any key pair argument like base url or anything by that user can use their own self hosted url??

@Navyasree-ulava  let me know??

File: sdks/urbackend-python/tests/__init__.py (hunk 1 above)
Comment (yash-pouranik) [reply sdks/urbackend-python/tests/__init__.py:1]: Okay

File: sdks/urbackend-python/src/client.py (hunk 2 above)
Comment (Navyasree-ulava) [reply sdks/urbackend-python/src/client.py]: actually i should keep it as an env variable so that user can user thier's self hosted or official url,  i updated env but not .env.example i should do that and replace that line of code which u mentioned

Comment (yash-pouranik) [reply sdks/urbackend-python/src/client.py]: Nooo
Not in env - Best is this - 
when user will import the sdk
then it should be like.

for defualt - `client = UrBackendClient(api_key="pk_live_YOUR_KEY")`
for self host - 
`client = UrBackendClient(api_key="pk_live_YOUR_KEY", base="https://domain.com")`
or
`client = UrBackendClient(api_key="pk_live_YOUR_KEY", base=env.base)`


This is the same approach I followed in js SDK
@Navyasree-ulava

Comment (Navyasree-ulava) [reply sdks/urbackend-python/src/client.py]: yeah it makes sense, i will also use the same approach

File: sdks/urbackend-python/tests/test_auth.py
Comment (Navyasree-ulava) [reply sdks/urbackend-python/tests/test_auth.py:1]: Screenshot 2026-04-26 162148

Comment (yash-pouranik) [reply sdks/urbackend-python/tests/test_auth.py:1]: okay.

**Review by yash-pouranik — APPROVED** [review]
Nicely Handeld

**Comment by yash-pouranik** [comment]
Hello @Navyasree-ulava  
Some more actually 👉👈

**Comment by Navyasree-ulava** [comment]
> Hello @Navyasree-ulava Some more actually 👉👈

resolving.........

**Review by yash-pouranik — APPROVED** [review]
@Navyasree-ulava 
okay thank you for the PR and Cooperating very well.
Merging now.
Hope to see more contributions from you.
Lastly Please leave a Star to the repo.

**Comment by Navyasree-ulava** [comment]
Thank u so much! 😊
Really enjoyed working on this and learned a lot through the process.

I’ve starred the repo ⭐
Looking forward to contributing more!
